File: solutions/0112/1.java
class Solution {
    boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }
        if (root.left == null && root.right == null) {
            return root.val == sum;
        }
        int target = sum - root.val;
        return hasPathSum(root.left, target)
            || hasPathSum(root.right, target);
    }
}

File: solutions/0112/1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return traverse(root, targetSum);
    }

    bool traverse(TreeNode* root, int target) {
        if (root == nullptr) {
            return false;
        }
        if (root->left == nullptr && root->right == nullptr) {
            return root->val == target;
        }
        bool left = traverse(root->left, target - root->val);
        bool right = traverse(root->right, target - root->val);
        return left || right;
    }
};
